### PR TITLE
Boxing logic reuses stack

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55575 Boxing logic reuses stack**

Previously, the boxing logic allocated a new stack for each call. Now, we reuse the stack.
Because we've seen slow performance of TLS in the past, this implementation doesn't use TLS
but a static variable, i.e. each operator signature stores an independent stack object.

This greatly improves perf of backend fallbacks.

I compared 4 experiments.
- **base:** The base case, no backend fallback involved. Dispatcher calls directly to the CPU kernel.
- **per-op codegen:** This is what XLA does today. The dispatcher calls into an op-specific unboxed backend kernel, that then calls into the dispatcher to redispatch to the CPU kernel.
- **fallthrough:** Use a generic backend fallback, but don't implement it manually, use a fallthrough instead.
- **backend fallback:** Implement a boxed backend fallback that is generic, i.e. works for all ops, and calls the dispatcher to redispatch to the CPU kernel.

I ran experiments on 2 operators: `aten::acos` and `aten::acos_out`.
- `aten::acos_out` somewhat underestimates the dispatch overhead because we don't actually unbox the result argument from the stack and instead just return a reference to the input argument for out overloads.
- `aten::acos` heavily overestimates the overhead because it internally calls `aten::empty` and that goes through the fallback too, so it's more an upper bound on 2x the dispatcher overhead.

Each experiment is run
- **before** this PR and
- **after** this PR

|  aten::acos_out  | instructions (before) | relative (before)  | instructions (after) | relative (after) |
|--------------------|---------------|--------------|--|--|
| base               | 3454021       |              | 3452021 | |
| per-op codegen     | 3508021       | base + 1.6%  | 3506021 | base + 1.6% |
| fallthrough        | 3454021       | base + 0%    | 3452021 | base + 0% |
| backend fallback   | 3932021       | base + 13.8% | 3713021 | base + 7.6% |

|  aten::acos      | instructions (before)  | relative (before)    | instructions (after) | relative (after) |
|--------------------|---------------|--------------|--|--|
| base               | 5196446       |              | 5208368 | |
| per-op codegen     | 5354446       | base + 3.0%  | 5366446 | base + 3% |
| fallthrough        | 5196446       | base + 0%    | 5208368 | base + 0% |
| backend fallback   | 7399466       | base + 42.4% | 6970190 | base + 33.8% |

In summary, we see that none of the cases regresses and the backend fallback overhead gets reduced from 13.8% to 7.6%.

Differential Revision: [D27638297](https://our.internmc.facebook.com/intern/diff/D27638297/)

### CODE REVIEW

**Comment by ezyang** [comment]
I'm a little skeptical about this. C++ vectors don't ever shrink so you could end up with a pretty wild memory leak due to this.

**Comment by smessmer** [comment]
> I'm a little skeptical about this. C++ vectors don't ever shrink so you could end up with a pretty wild memory leak due to this.

They not shrinking is exactly what I'm relying on here. On the first call, they'll grow to the number of arguments, and after the first call they'll never change size again.

**Comment by swolchok** [comment]
> this implementation doesn't use TLS but a static variable, i.e. each operator signature stores an independent stack object.

I am confused. How can this be thread-safe? Boxing mutates the stack, the static will be shared between threads, and there shouldn't be anything stopping multiple threads from running operators.

**Comment by smessmer** [comment]
@swolchok oops you're right. That's why we were thinking about TLS when we thought about this optimization the first time, lol

**Comment by ezyang** [comment]
What makes more sense to me is adjusting the calling convention of boxed operators so that they take in a stack as argument, which they can use for redispatch.

**Comment by smessmer** [comment]
@ezyang that might work. Note though that you'd have to change the calling convention of **unboxed** operators to somehow pass through a stack. Boxed operators already do that, but the stack isn't passed through unboxed operators. That's why it currently gets re-created every time you go back from unboxed to boxed.

Since unboxed operators don't really do anything with a stack, it would be just passed in in case they call a boxed operator and otherwise ignored. Seems like an unnecessary API hindrance, but we actually have use cases for passing in OperatorHandle and DispatchKeySet to unboxed operators too. Maybe it's time to combine those things up into a CallContext struct and just pass that in.

**Comment by ezyang** [comment]
Hmm, never mind, you're right, I don't want you to pass in a stack in this case. Looks like you gonna have to do some TLS. Fortunately boxing is already somewhat slow so maybe it won't matter that much.
